### DISCUSSION TITLE
Using a Language Server, an extension typically needs to analyze documents which are not open (e.g. to discover declarations included in other files). How to do it?

### DISCUSSION THREAD

**Discussion by pl7y** [microsoft/vscode-discussions#633] [Extension Development QnA] [answered]
In a language extension, you might have two files `one.mylang` and `two.mylang` in which the former includes the latter (e.g. `one.mylang` uses a variable declared in `two.mylang`. So for validation, completion, highlight and so on we need to **parse all files in the workspace,** not only those that are open.

The [LSP guide](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide) does not mention it and for the sake of me I couldn't find anything related 🙄, apart from [very](https://github.com/microsoft/vscode/issues/15723) [old](https://github.com/microsoft/vscode-languageserver-node/issues/31) [hints](https://github.com/microsoft/language-server-protocol/issues/837) which don't make clear how to do this (apart from stating that the LSP is in charge of that parsing).

From how I see it, the point is that in the LSP the TextDocuments object has only files opened in the client; do we need to forcibly add non-opened files to it (somehow)? Or have another list of unopened documents (but, in this case)? Or does the vscode LSP implementation offer some functionality for it?

Thank you so much, I'm pretty stuck with that.

**Comment by r3m0t** [comment]
Hi, the URI of the workspace is in [InitializeParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeParams) as `rootUri` and `workspaceFolders[].uri` , if the scheme is `'file'` you can read all files from the filesystem (unless they are open).

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didOpen

> The document’s content is now managed by the client and the server must not try to read the document’s content using the document’s Uri... Note that a server’s ability to fulfill requests is independent of whether a text document is open or closed.

As a helper you can use this package to deal with URI-to-path on all platforms- https://github.com/microsoft/vscode-uri

If the scheme is not `file` then you need to implement the logic of the scheme in your LSP, or follow this issue - https://github.com/microsoft/language-server-protocol/issues/1264

**Comment by pl7y** [comment]
Ok, so if I got your answer correctly, I try to recap:
* let's assume scheme is `file:`;
* the LSP needs to `fs.readFile` all the files in the `workspaceFolders` or `rootUri`: 

```ts
		this.connection.onInitialize((params: InitializeParams) => {
			const capabilities = params.capabilities;
			if (params.workspaceFolders) {
				const workspaceFolder = params.workspaceFolders[0];
                                readAllFilesIn(workspaceFolder);
			}
```
* with the files read, the LSP must manage all parsing, validation and so on. Correct?

**Reply by r3m0t** [reply, marked as answer]
Pretty much. It's more efficient to only read the files that are relevant. Most languages have `import`/`require`/`using` statements so that you don't have to read every file.

For each file you read you also need to:

* Watch them for changes - the client can tell you when a file has changed, you then need to reread the file. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles
* Update them from the editor - when you get `textDocument/didOpen`, you start using the version of the document in the editor, when it's closed you read it from the filesystem.

If your LSP is written in JS and you use `vscode-languageserver-textdocument` it can handle a lot of this for you - https://code.visualstudio.com/api/language-extensions/language-server-extension-guide

**Reply by pl7y** [reply]
This figures, thank you so much!
